src/nbl/ui/CWindowWin32.cpp

```diff
@@ -5,6 +5,218 @@
+
+			int32_t windowLeft = (GetSystemMetrics(SM_CXSCREEN) - realWidth) / 2;
+			int32_t windowTop = (GetSystemMetrics(SM_CYSCREEN) - realHeight) / 2;
+
+			if (windowLeft < 0)
+				windowLeft = 0;
+			if (windowTop < 0)
+				windowTop = 0;	// make sure window menus are in screen on creation
+
+			if (m_params.flags & CWindowWin32::ECF_FULLSCREEN)
+			{
+				windowLeft = 0;
+				windowTop = 0;
+			}
+
+			// create window
+
+			m_params.nativeWindow = CreateWindow(classname, __TEXT(""), style, windowLeft, windowTop,
+				realWidth, realHeight, NULL, NULL, hinstance, NULL);
+			if ((m_params.flags & CWindowWin32::ECF_HIDDEN) == 0)
+				ShowWindow(m_params.nativeWindow, SW_SHOWNORMAL);
+			UpdateWindow(m_params.nativeWindow);
+
+			// fix ugly ATI driver bugs. Thanks to ariaci
+			// TODO still needed?
+			MoveWindow(m_params.nativeWindow, windowLeft, windowTop, realWidth, realHeight, TRUE);
+		}
+		void setParams(SParams&& params) override { m_params = static_cast<SCreateWindowParams&&>(params); }
+	};
+	struct SDestroyWindowRequest : SRequest
+	{
+		SDestroyWindowRequest(CWindowWin32::native_handle_t wnd) : m_params{ wnd } {}
+		struct SDestroyWindowParams : SParams
+		{
+			CWindowWin32::native_handle_t window;
+		} m_params;
+		void process() override
+		{
+			DestroyWindow(m_params.window);
+		}
+		void setParams(SParams&& params) override { m_params = static_cast<SDestroyWindowParams&&>(params); }
+	};
+	class CThreadHandler final : public system::IAsyncQueueDispatcher<CThreadHandler, SRequest, 256u>
+	{
+		using base_t = system::IAsyncQueueDispatcher<CThreadHandler, SRequest, 256u>;
+		friend base_t;
+	public:
+		void createWindow(int _x, int _y, uint32_t _w, uint32_t _h, CWindowWin32::E_CREATE_FLAGS _flags, CWindowWin32::native_handle_t wnd)
+		{
+			SCreateWindowRequest rq(_x, _y, _w, _h, _flags, wnd);
+			request(rq);
+			waitForCompletion(rq);
+		}
+		void destroyWindow(CWindowWin32::native_handle_t window)
+		{
+			SDestroyWindowRequest rq(window);
+			request(rq);
+			waitForCompletion(rq);
+		}
+		CThreadHandler()
+		{
+			this->start();
+		}
+
+	private:
+		void waitForCompletion(SRequest& req)
+		{
+			auto lk = req.wait();
+		}
+
+	private:
+		void init() {}
+
+		void exit() {}
+
+		void background_work()
+		{
+			static MSG message;
+			static uint32_t timeoutInMS = 8; // gonna become 10 anyway
+			if (getMessageWithTimeout(&message, timeoutInMS))
+			{
+				TranslateMessage(&message);
+				DispatchMessage(&message);
+			}
+		}
+
+		void process_request(SRequest& req)
+		{
+			req.process();
+		}
+
+		template <typename RequestParams>
+		void request_impl(SRequest& req, RequestParams&& params)
+		{
+			req.setParams(std::move(params));
+		}
+	private:
+		static bool getMessageWithTimeout(MSG* msg, uint32_t timeoutInMilliseconds)
+		{
+			bool res;
+			UINT_PTR timerId = SetTimer(NULL, NULL, timeoutInMilliseconds, NULL);
+			res = GetMessage(msg, nullptr, 0, 0);
+			
+			PostMessage(nullptr, WM_NULL, 0, 0);
```

Comment:
what is this PostMessage for?

Reply:
https://github.com/Crisspl/Nabla/pull/5#discussion_r641950432, if i understood that suggestion correctly

Reply:
Nah, i guess i didn't quite understand that

Reply:
yeah, the waking thread posts the message